modules/nf-core/samtools/pipeline/main.nf

```diff
@@ -0,0 +1,112 @@
+    def n_commands = commands.size()
+    if (n_commands < 2) error "SAMTOOLS_PIPELINE is used to chain 2 or more samtools commands"
+
+    // Fetch the arguments
+    def all_args = []
+    for (int index = 0; index < n_commands; index++) {
+        all_args.add( task.ext.args?[ commands[index] ] ?: '' )
+    }
+    def last_args = all_args[-1]
+
+    // Output file
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def extension = last_args.contains("--output-fmt sam") ? "sam" :
+                    last_args.contains("--output-fmt bam") ? "bam" :
+                    last_args.contains("--output-fmt cram") ? "cram" :
+                    "bam"
+    if ("$input" == "${prefix}.${extension}") error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
+
+    def pipeline_command = ""
+    for (int index = 0; index < n_commands; index++) {
+        def this_command = commands[index]
+        def is_first_command = (index == 0)
+        def is_last_command = (index == (n_commands - 1))
+
+        pipeline_command += """
+        samtools ${this_command} ${all_args[index]} \\
+        """
+
+        // The reheader has no useful option
+        if (! ["reheader"].contains(this_command)) {
+            pipeline_command += " -@ $task.cpus"
+            if (is_last_command) {
+                pipeline_command += fasta ? " --reference ${fasta}" : ""
+            } else {
+                pipeline_command += " -u"
+            }
+        }
+        this_input = (is_first_command ? " $input" : " -")
+
+        // samtools commands have slightly different syntax
+        if (["collate"].contains(this_command)) {
+            // [-o OUTPUT|-O] [INPUT|-]
+            pipeline_command += is_last_command ? " -o ${prefix}.${extension}" : " -O"
+            pipeline_command += this_input
+
+        } else if (["addreplacerg", "sort", "view"].contains(this_command)) {
+            // [-o OUTPUT] [INPUT|-]
+            pipeline_command += is_last_command ? " -o ${prefix}.${extension}" : ""
+            pipeline_command += this_input
+
+        } else if (["reheader"].contains(this_command)) {
+            // [INPUT|-]
+            pipeline_command += this_input
+
+        } else if (["fixmate", "markdup"].contains(this_command)) {
+            // [INPUT|-] [OUTPUT|-]
+            pipeline_command += this_input
+            pipeline_command += is_last_command ? " ${prefix}.${extension}" : " -"
+
+        } else {
+            error "${this_command} is not supported"
+        }
+
+        if (!is_last_command) {
+            pipeline_command += " |"
+        }
+    }
+
+    """
+    $pipeline_command
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
+    END_VERSIONS
+    """
+
+    stub:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    touch ${prefix}.${extension}
```

Comment:
Not really what you wanted comments on, but this stub won't work since the `extension` isn't defined in the stub scope (all output files will be `${prefix}.null`)

Reply:
I noticed that too :)